tests/apollo/test_skvbc_backup_restore.py

```diff
@@ -0,0 +1,134 @@
+# You may not use this product except in compliance with the Apache 2.0 License.
+#
+# This product may include a number of subcomponents with separate copyright
+# notices and license terms. Your use of these subcomponents is subject to the
+# terms and conditions of the subcomponent's license, as noted in the LICENSE
+# file.
+import os.path
+import random
+import unittest
+
+import trio
+
+from util import skvbc as kvbc
+from util.bft import with_trio, with_bft_network, with_constant_load, KEY_FILE_PREFIX
+
+
+def start_replica_cmd(builddir, replica_id):
+    """
+    Return a command that starts an skvbc replica when passed to
+    subprocess.Popen.
+
+    The replica is started with a short view change timeout and with RocksDB
+    persistence enabled (-p).
+
+    Note each arguments is an element in a list.
+    """
+    statusTimerMilli = "500"
+    viewChangeTimeoutMilli = "10000"
+    path = os.path.join(builddir, "tests", "simpleKVBC", "TesterReplica", "skvbc_replica")
+    return [path,
+            "-k", KEY_FILE_PREFIX,
+            "-i", str(replica_id),
+            "-s", statusTimerMilli,
+            "-v", viewChangeTimeoutMilli,
+            "-p",
+            "-t", os.environ.get('STORAGE_TYPE')]
+
+
+class SkvbcBackupRestoreTest(unittest.TestCase):
+
+    __test__ = False  # so that PyTest ignores this test scenario
+
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    async def test_checkpoint_propagation_after_restarting_replicas(self, bft_network):
+        """
+        Here we trigger a checkpoint, restart all replicas in a random order with 10s delay in-between,
+        both while stopping and starting. We verify checkpoint persisted upon restart and then trigger
+        another checkpoint. We make sure checkpoint is propagated to all the replicas.
+
+        1) Given a BFT network, we make sure all nodes are up
+        2) Send sufficient number of client requests to trigger checkpoint protocol
+        3) Stop all replicas in a random order (with 10s delay in between)
+        4) Start all replicas in a random order (with 10s delay in between)
+        5) Send sufficient number of client requests to trigger another checkpoint
+        6) Make sure checkpoint propagates to all the replicas
+        7) Make sure the initial view is stable
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        n = bft_network.config.n
+
+        self.assertEqual(len(bft_network.procs), n, "Make sure all replicas are up initially.")
+
+        current_primary = await bft_network.get_current_primary()
+
+        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=current_primary)
+
+        await skvbc.fill_and_wait_for_checkpoint(
+            initial_nodes=bft_network.all_replicas(),
+            checkpoint_num=1,
+            verify_checkpoint_persistency=False
+        )
+
+        # verify checkpoint propagation to all the nodes
```

Comment:
I don't think you need this. [This call](https://github.com/vmware/concord-bft/blob/master/tests/apollo/util/skvbc.py#L229), made as part of `skvbc.fill_and_wait_for_checkpoint`, already waits for all replicas to reach the checkpoint.

Reply:
Thanks @andrewjstone for the review. Yes, that's true. The reason I have done it this way is to verify checkpoint creation using assertEqual in the loop.

Reply:
You won't even get to this point unless that condition is true though.

Reply:
`wait_for_replicas_to_checkpoint` does not take `expected_checkpoint_num` as argument. Hence, the underlying call to `wait_for_checkpoint` can return with only `last_stored_checkpoint`, without waiting for replica to reach the expected checkpoint number.

Reply:
@andrewjstone, updated as discussed.